Make requesting user available in context

liggitt opened the pull request:
This PR does the following:
- Remove the map of requests to users
- Add a map of requests to contexts
- Initializes a context containing the user as the first step of an authenticated request
- Pass the requestToContext map into resthandler
- Modify resthandler to use the context already associated with the request if one exists

This feels gorpy, but I can't think of a better way to let resthandler have requesting user info without a global request-to-context (or user) map. Open to suggestions for cleaner ways to accomplish this.

@smarterclayton @erictune

liggitt committed:
Add user helper methods to context

liggitt committed:
Put user in context, map requests to context above resthandler layer

smarterclayton commented:
@erictune FYI as well - this establishes user as part of context, which allows subordinate calls to tie a request back to a user.

liggitt commented on pkg/api/requestcontext.go:
```diff
@@ -0,0 +1,92 @@
+	return nil
+}
+
+func (c *requestContextMap) set(req *http.Request, context Context) {
```
This is me trying to prevent the world from monkeying with the internal map

liggitt commented on pkg/api/requestcontext.go:
```diff
@@ -0,0 +1,92 @@
+	return context, ok
+}
+
+func (c *requestContextMap) Update(req *http.Request, context Context) error {
```
And yet, to allow the auth layer to update the context for the request to one with the user value set, something has to let us update the map (but only if it exists, so we can be sure it will get cleaned up?)

smarterclayton replied:
I think so.

----- Original Message -----
> > +}
> > +
> > +func NewRequestContextMapper() RequestContextMapper {
> > +	return &requestContextMap{
> > +		contexts: make(map[*http.Request]Context),
> > +	}
> > +}
> > +
> > +func (c *requestContextMap) Get(req *http.Request) (Context, bool) {
> > +	c.lock.Lock()
> > +	defer c.lock.Unlock()
> > +	context, ok := c.contexts[req]
> > +	return context, ok
> > +}
> > +
> > +func (c *requestContextMap) Update(req *http.Request, context Context)
> > error {
> 
> And yet, to allow the auth layer to update the context for the request to one
> with the user value set, something has to let us update the map (but only if
> it exists, so we can be sure it will get cleaned up?)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4352/files#r24539620

smarterclayton commented:
In my head there are three steps that happen during every request:

1. An initial context is established in a global map (req -> context)
2. A user is authenticated
3. A user is authorized

2 and 3 are subordinate to 1.  Code below 3 should have the api.Context available and try to pass it down, and generally should not have to modify the api.Context (other than through the standard mechanism of scoping it and passing it to a child).

liggitt commented:
@derekwaynecarr FYI

smarterclayton commented on pkg/api/requestcontext.go:
```diff
@@ -0,0 +1,92 @@
+	"sync"
+)
+
+type RequestContextMapper interface {
```
Godoc

smarterclayton replied:
Describe who should be allowed to call update and what the risks are

smarterclayton commented on pkg/api/requestcontext.go:
```diff
@@ -0,0 +1,92 @@
+func (c *requestContextMap) Update(req *http.Request, context Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, ok := c.contexts[req]; !ok {
```
Passing a nil new context should be an error.  I don't think we need to allow clearing.

Eventually, we may want to verify this is a descendant context and fail if not.  Add a TODO (unless that's trivial today)

smarterclayton commented on pkg/api/requestcontext.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// Returns true if there are no contexts registered. Intended for use by tests.
+func IsEmpty(requestsToContexts RequestContextMapper) (bool, error) {
```
Godoc style (IsEmpty returns true ...)

smarterclayton commented on pkg/auth/handlers/handlers.go:
```diff
@@ -18,39 +18,36 @@ package handlers
-				glog.Errorf("Unable to authenticate the request due to an error: %v", err)
+func NewRequestAuthenticator(requestContextMapper api.RequestContextMapper, auth authenticator.Request, failed http.Handler, handler http.Handler) http.Handler {
+	return api.NewRequestContextFilter(
+		requestContextMapper,
```
Nit: call it mapper

smarterclayton commented:
Minor comments, LGTM otherwise.  In order to use context as its intended we would have needed to do this anyway.

smarterclayton commented:
The only person who should update would be the authenticator or the tracer (my other WIP).  Everyone else should be passing context down.

liggitt commented:
updated:
* addressed comments
* added godoc
* renamed `set` to `init`
* made sure an initial context will only get set once
* made sure `remove` is only paired with the first successful `init`
* fail early in `NewRequestContextFilter` if an unknown mapper type is given

liggitt commented:
rebased, added a ContextFunc for api_installer to use

smarterclayton commented on pkg/master/master.go:
```diff
@@ -367,8 +372,10 @@ func logStackOnRecover(panicReason interface{}, httpWriter http.ResponseWriter)
 func (m *Master) init(c *Config) {
-	var userContexts = handlers.NewUserRequestContext()
-	var authenticator = c.Authenticator
+	var err error
```
Would probably move this out.

smarterclayton commented on pkg/master/master.go:
```diff
@@ -367,8 +372,10 @@ func logStackOnRecover(panicReason interface{}, httpWriter http.ResponseWriter)
-	var authenticator = c.Authenticator
+	var err error
+	if m.requestContextMapper == nil {
+		m.requestContextMapper = api.NewRequestContextMapper()
```
I think this should be in setDefaults

liggitt replied:
done

smarterclayton commented on pkg/master/master.go:
```diff
@@ -89,6 +89,9 @@ type Config struct {
 	MasterServiceNamespace string
 
+	// Map requests to contexts
+	RequestContextMapper api.RequestContextMapper
```
Add a comment that this is exposed so that downstream projects can provide their own mappers

liggitt replied:
done

smarterclayton commented on pkg/master/master.go:
```diff
@@ -471,6 +481,16 @@ func (m *Master) init(c *Config) {
 	}
 
+	// After all wrapping is done, put a context filter around both handlers
+	m.Handler, err = api.NewRequestContextFilter(m.requestContextMapper, m.Handler)
```
```
handler, err := api.NewRequestContextFilter
if err != nil {
}
m.Handler = handler
```

liggitt replied:
scoped errors locally

smarterclayton commented on pkg/master/master.go:
```diff
@@ -453,12 +460,15 @@ func (m *Master) init(c *Config) {
-	if authenticator != nil {
-		handler = handlers.NewRequestAuthenticator(userContexts, authenticator, handlers.Unauthorized, handler)
+	if c.Authenticator != nil {
+		handler, err = handlers.NewRequestAuthenticator(m.requestContextMapper, c.Authenticator, handlers.Unauthorized, handler)
```
I don't like not scoping errs to the context they're called in.

liggitt replied:
fixed

liggitt commented:
rebased and comments addressed

smarterclayton commented:
This LGTM but want @erictune's signoff

smarterclayton commented:
Got impatient, will take feedback from master :)
